### DISCUSSION TITLE
Aux and Chamber Fan support

### DISCUSSION THREAD

**Discussion by lettore** [jschuh/klipper-macros#110] [Ideas]
I added to my printer an Auxiliary and a Chamber Fan like Bambu X1C have, but also Vorons.
Orcaslicer uses the standard Marlin way of controlling the fan with M106 Px Sx command.
So M106 P1 or M106 control the part fan
M106 P2 the aux fan
M106 P3 the chamber fan
The same does M107.
I did the coding but I'm not sure if can be useful to add support for scaling all the fans.
I also added the LCD menu entry for changing manually the aux fan speed as the slicer turns on the fan at print start-up and turns it off only at the print end.
Personally I have a different use of the Chamber fan as it's a temperature fan set to +2 degrees of the Chamber heater temp, so it turns on only if the temperature is above the requested one.

**Comment by jschuh** [comment]
Yeah, I thought about supporting an arbitrary number of fans when I wrote the fan wrappers a few years ago, but decided against it since at the time I wasn't aware of a common convention for associating the fan config with its index. If the convention is now to map aux and chamber as P1 and P2, then I'd be fine with extending the scaling to arbitrary fans. It's easy enough to do after all.

**Comment by lettore** [comment]
Well is not a convention, the only slicer that manages the Aux Fan and Chamber Fan is Orcaslicer for Bambu printers, this is the way the Bambulabs implemented this.
I opened a PR, I'm not really a coder in Git so please forgive me for any error.
I don't think that scaling is useful for extra fans, at least for Chamber but also for Aux as the slicer only emits the Gcode at the print start, it's possible to tune the speed from the interface or the WebUI. Maybe it will be more useful an option to "link" the speed of the Aux Fan to the Part Fan, so it can be used as an addition to the Part Fan and controlled by the same command M106.

**Comment by jschuh** [comment]
I appreciate the effort, but that's not the kind of change I would accept. As I mentioned in my earlier comment, I can see a rationale for extending fan scaling to arbitrary `fan_generic` configs, and as part of that also establishing a naming convention to expose those fans to the `M106`/`M107` commands via the `P` index parameter (e.g. for idex printers).

However, #112 just maps two fan names to fixed indexes, which I don't see as generic enough to land in the project. It's more of a niche slicer customization that I think is better handled with small wrapper macros in the local printer.cfg. Here's an example of how to do it:

```
[gcode_macro m106]
rename_existing: M106.1
gcode:
  {% set S = params.S|default(255)|int %}
  {% set P = params.P|default(1)|int %}
  {% if P == 1 %}
    M106.1 {rawparams}
  {% elif P == 2 %}
    SET_FAN_SPEED FAN=aux_fan SPEED={S/255.0}
  {% elif P == 3 %}
    SET_FAN_SPEED FAN=chamber_fan SPEED={S/255.0}
  {% endif %}

[gcode_macro m107]
rename_existing: M107.1
gcode:
  {% if params.P|default(1)|int == 1 %}
    M107.1
  {% else %}
    M106 S0 {rawparams}
  {% endif %}
```

**Comment by lettore** [comment]
I don't know who use more than these fans and how you can drive them if no slicer supports them but I got your point.
